### PR TITLE
Add support for user specific hardware and disk

### PR DESCRIPTION
**Pull request by phracek**
Add support for user specific hardware.

In case hardware is properly specified as JSON string,
like '{"memory": ">= 8 GB"}', then new method called
api.unsafeNewRequest is called.

In case hardware is not specified or it is proper JSON format
then api.newRequest is called as before.

action.yml is updated with support 'hardware' option.

For more information about hardware support see https://tmt.readthedocs.io/en/latest/spec/hardware.html

Tests added for this pull request are:
test/action.test.ts: 'Hardware test', and 'Hardware test - wrong JSON'
test/unit/unit.test.ts: tests for isJsonString

Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

For more information see:
https://gitlab.com/testing-farm/docs/root/-/merge_requests/120/diffs?view=inline
https://api.testing-farm.io/redoc#operation/request_a_new_test_v0_1_requests_post

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -42,12 +42,13 @@ See [Testing Farm onboarding guide](https://docs.testing-farm.io/Testing%20Farm/
 
 ### Test Environment
 
-| Input Name | Description | Default value |
-|------------|-------------|---------------|
-| `compose` | Compose to run tests on. [Available composes.](https://api.dev.testing-farm.io/v0.1/composes) | Fedora-latest |
-| `arch` | Define an architecture for testing environment | x86_64 |
-| `variables` | Environment variables for test env, separated by ; | empty |
-| `secrets` | Environment secrets for test env, separated by ; | empty |
+| Input Name  | Description                                                                                    | Default value |
```
Comment (zmiklank) [inline README.md:45]: Please do not make all the rows the same length. The git history is then harder to read.

Comment (phracek) [reply README.md]: Fixed by https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/3d807fe8f0e1940a6d7dc4901641eb18e190f868

File: src/action.ts [hunk 2]
```diff
@@ -113,6 +134,8 @@ async function action(pr: PullRequest): Promise<void> {
         settings: envSettings,
         secrets: tmtEnvSecrets,
         artifacts: tmtArtifacts,
+        hardware: tmtHardware,
+        disk: tmtDisk,
```
Comment (thrix) [inline src/action.ts:138]: disk is part of the HW?

File: src/schema/input.ts [hunk 3]
```diff
@@ -34,6 +34,37 @@ export const tmtArtifactsInputSchema = stringToArraySchema.transform(arr => {
 });
 export const tmtContextInputSchema = keyValueArrayToObjectSchema;
 
+// See https://tmt.readthedocs.io/en/stable/spec/plans.html#hardware
+// https://gitlab.com/testing-farm/docs/root/-/merge_requests/120/diffs?view=inline
+export const tmtHardwareSchema = z
```
Comment (thrix) [inline src/schema/input.ts:39]: This is definitely not complete, I wonder if we need to validate it here, if so, we should provide rather some npm liberary for validation:

https://gitlab.com/testing-farm/artemis/-/blob/main/server/src/tft/artemis/schema/environment-v0.0.69.yml?ref_type=heads

Could we for now just use an object, and not validate it?

**Review by thrix — CHANGES_REQUESTED** [review]

File: src/action.ts (hunk 2 above)
Comment (phracek) [reply src/action.ts:138]: > disk is part of the HW?

AFAIK not. based on the link https://gitlab.com/testing-farm/docs/root/-/merge_requests/120/diffs?view=inline

File: src/schema/input.ts (hunk 3 above)
Comment (phracek) [reply src/schema/input.ts:39]: I will have a look at it with team. thanks for suggestion. I need to find out solution how to implement it.

File: src/action.ts (hunk 2 above)
Comment (phracek) [reply src/action.ts]: Removed by commit https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/bddea9b2ace8b449d665c9889534bdb6b2497738

File: src/schema/input.ts (hunk 3 above)
Comment (phracek) [reply src/schema/input.ts]: It should be fixed by https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/61532115af926ad0d9a5676858afd0c4e7ebd3d1

**Comment by phracek** [comment]
This pull request is blocked by https://github.com/redhat-plumbers-in-action/testing-farm/pull/111

File: README.md [hunk 4]
```diff
@@ -48,6 +48,7 @@ See [Testing Farm onboarding guide](https://docs.testing-farm.io/Testing%20Farm/
 | `arch` | Define an architecture for testing environment | x86_64 |
 | `variables` | Environment variables for test env, separated by ; | empty |
 | `secrets` | Environment secrets for test env, separated by ; | empty |
+| `hardware` | Hardware definition for test env | empty |
```
Comment (jamacku) [inline README.md:51]:
```suggestion
| `tmt_hardware` | Hardware definition for test env | empty |
```

File: action.yml [hunk 5]
```diff
@@ -83,6 +83,10 @@ inputs:
     description: 'Action is waiting for testing farm to finish or until timeout is reached. Value is in minutes.'
     required: false
     default: '480'
+  hardware:
```
Comment (jamacku) [inline action.yml:86]:
```suggestion
  tmt_hardware:
```

**Review by jamacku — CHANGES_REQUESTED** [review]

File: src/util.ts [hunk 6]
```diff
@@ -20,3 +20,12 @@ export function getSummary(result: unknown): string {
 
   return parsedResult.success ? ` - ${parsedResult.data.summary}` : '';
 }
+
+export function isJsonString(hardware: string): boolean {
+  try {
+    JSON.parse(hardware);
+  } catch (e) {
+    return false;
```
Comment (jamacku) [inline src/util.ts:28]: @phracek Maybe add some logging?

Comment (jamacku) [reply src/util.ts:28]: Wouldn't it be better not to catch the exception and let the action fail? If somebody is trying to set hardware requirements, it's probably required for the test to run successfully.

**Comment by zmiklank** [comment]
@phracek Thanks for the contribution, I did not do a review, I believe jamacku will cover everything needed. But as I mentioned I am linking our [integration tests](https://github.com/sclorg/testing-farm-as-github-action/tree/main/.github/workflows). Feel free to add some to cover this new feature too. You can take an inspiration for example from [github summary test](https://github.com/sclorg/testing-farm-as-github-action/blob/main/.github/workflows/github_summary_test.yml) and its tmt plan.

File: src/util.ts (hunk 6 above)
Comment (phracek) [reply src/util.ts]: This is fixed by commit https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/cb3b7bc435c80b0563d52914ba7235c4985b6d1f. 
Function isJsonString is not needed now.

File: action.yml [hunk 7]
```diff
@@ -83,6 +83,10 @@ inputs:
     description: 'Action is waiting for testing farm to finish or until timeout is reached. Value is in minutes.'
     required: false
     default: '480'
+  tmt_hardware:
+    description: 'Define hardware requirements for the testing environment as a JSON string. See https://tmt.readthedocs.io/en/latest/spec/hardware.html'
+    required: false
+    default: '{}'
```
Comment (jamacku) [inline action.yml:89]: You probably don't want to set a default value. That would mean that every time unsafeNewRequest is used.

```suggestion
```

File: src/action.ts [hunk 8]
```diff
@@ -71,6 +71,15 @@ async function action(pr: PullRequest): Promise<void> {
     ? tmtArtifactsSchema.parse(tmtArtifactsParsed.data)
     : [];
 
+  // Generate tmt hardware specification
+  // See https://tmt.readthedocs.io/en/stable/spec/plans.html#hardware
+  // https://gitlab.com/testing-farm/docs/root/-/merge_requests/120/diffs?view=inline
+
+  let tmtHardware;
+  if (getInput('tmt_hardware') !== '') {
```
Comment (jamacku) [inline src/action.ts:79]: I would rather store `tmt_hardware` in constant and use that for comparison.

```suggestion
  const rawTmtHardware = getInput('tmt_hardware');
  if (rawTmtHardware) {
```

**Comment by phracek** [comment]
[test]

File: src/action.ts [hunk 9]
```diff
@@ -120,8 +129,17 @@ async function action(pr: PullRequest): Promise<void> {
     ],
   };
 
-  // The strict mode should be enabled once https://github.com/redhat-plumbers-in-action/testing-farm/issues/71 is fixed
-  const tfResponseRaw = await api.newRequest(request, false);
+  let tfResponseRaw;
+  // Use newRequest method in case tmt_hardware is not defined or parameter is not properly formatted
+  if (getInput('tmt_hardware') === '') {
```
Comment (jamacku) [inline src/action.ts:134]:
```suggestion
  if (!rawTmtHardware) {
```

**Review by jamacku — CHANGES_REQUESTED** [review]

**Comment by phracek** [comment]
Added distribution sources

[test]

File: src/action.ts [hunk 10]
```diff
@@ -71,6 +71,16 @@ async function action(pr: PullRequest): Promise<void> {
...
     : [];
 
+  // Generate tmt hardware specification
+  // See https://tmt.readthedocs.io/en/stable/spec/plans.html#hardware
+  // https://gitlab.com/testing-farm/docs/root/-/merge_requests/120/diffs?view=inline
+
+  let tmtHardware;
+  const rawTmtHardware = getInput('tmt_hardware');
+  if (rawTmtHardware) {
+    tmtHardware = JSON.parse(getInput('tmt_hardware'));
```
Comment (jamacku) [inline src/action.ts:81]:
```suggestion
    tmtHardware = JSON.parse(rawTmtHardware);
```

File: tests/unit/util.test.ts [hunk 11]
```diff
@@ -1,6 +1,10 @@
 import { test, expect, describe } from 'vitest';
 
-import { composeStatusDescription, getSummary } from '../../src/util';
+import {
+  composeStatusDescription,
+  getSummary,
+  isJsonString,
```
Comment (jamacku) [inline tests/unit/util.test.ts:6]:
```suggestion
```

**Review by jamacku — CHANGES_REQUESTED** [review]

File: tests/action.test.ts [hunk 12]
```diff
@@ -245,6 +247,125 @@ describe('Integration tests - action.ts', () => {
...
+        state: 'complete',
+        result: { overall: 'passed', summary: '\\o/' },
+      });
+
+    // Run action
+    const octokit = new Octokit({ auth: 'mock-token' });
+    const pr = await PullRequest.initialize(1, octokit);
+
+    await action(pr);
+
```
Comment (jamacku) [inline tests/action.test.ts:287]: Maybe also test that `newReqeust()` wasn't called?

File: tests/action.test.ts [hunk 13]
```diff
@@ -245,6 +247,125 @@ describe('Integration tests - action.ts', () => {
     expect(mocks.TFError).not.toHaveBeenCalled();
   });
 
+  test('Hardware test', async () => {
+    setDefaultInputs();
+    // Mock required Action inputs
```
Comment (jamacku) [inline tests/action.test.ts:252]:
```suggestion

    // Mock required Action inputs
```

File: tests/action.test.ts [hunk 14]
```diff
@@ -245,6 +247,125 @@ describe('Integration tests - action.ts', () => {
...
+      await action(pr);
+    } catch (error) {
+      expect(error).toMatchInlineSnapshot(`
+        [SyntaxError: Unexpected token '}', "{"memory":}" is not valid JSON]
+      `);
+    }
+
+    // Test summary
+    // When timeout is reached, the summary is currently empty
+    await assertSummary(``);
```
Comment (jamacku) [inline tests/action.test.ts:366]: This is probably not needed. Please remove assert for empty summary.

**Review by jamacku — CHANGES_REQUESTED** [review]
Please also add assert to the minimal integration test that `unsafeNewRequest` wasn't called

File: tests/action.test.ts (hunk 14 above)
Comment (phracek) [reply tests/action.test.ts]: Fixed by https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/fb43f94db0bde185de3d4635938e3c90baf6851f

File: tests/action.test.ts (hunk 13 above)
Comment (phracek) [reply tests/action.test.ts:253]: Fixed by https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/fb43f94db0bde185de3d4635938e3c90baf6851f

File: tests/action.test.ts (hunk 12 above)
Comment (phracek) [reply tests/action.test.ts:288]: Fixed by https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/fb43f94db0bde185de3d4635938e3c90baf6851f

File: tests/action.test.ts [hunk 15]
```diff
@@ -245,6 +247,128 @@ describe('Integration tests - action.ts', () => {
...
+    vi.stubEnv(
+      'INPUT_GIT_URL',
+      'https://github.com/sclorg/testing-farm-as-github-action'
+    );
+    // tmt_plan_regex - A tmt plan regex which will be used for selecting plans. By default all plans are selected
+    vi.stubEnv('INPUT_TMT_PLAN_REGEX', 'fedora');
+    vi.stubEnv('INPUT_TMT_HARDWARE', '{"memory":}');
+
+    // Mock Testing Farm API
+    vi.mocked(mocks.newRequest).mockImplementation(
```
Comment (jamacku) [inline tests/action.test.ts:339]: Please remove mocks that are not used in this test to make it shorter. e.g. newRequest isn't used

**Review by jamacku — CHANGES_REQUESTED** [review]

File: tests/action.test.ts (hunk 15 above)
Comment (phracek) [reply tests/action.test.ts]: https://github.com/sclorg/testing-farm-as-github-action/pull/146/commits/1943d65d8ca8c50ca249f2e51ba590c6bf801785

**Review by jamacku — CHANGES_REQUESTED** [review]
> Please also add assert to the minimal integration test that `unsafeNewRequest` wasn't called

File: tests/action.test.ts (hunk 15 above)
Comment (jamacku) [reply tests/action.test.ts]: @phracek please also remove the rest of the mocks

File: tests/action.test.ts [hunk 16]
```diff
@@ -214,6 +216,80 @@ describe('Integration tests - action.ts', () => {
 
     await action(pr);
 
+    // hardware reguest has not been called
+    expect(mocks.unsafeNewRequest).not.toHaveBeenCalled();
+    // Check if we have waited for Testing Farm to finish
```
Comment (jamacku) [inline tests/action.test.ts:221]:
```suggestion

    // Check if we have waited for Testing Farm to finish
```

File: tests/action.test.ts [hunk 17]
```diff
@@ -214,6 +216,80 @@ describe('Integration tests - action.ts', () => {
...
+    expect(mocks.requestDetails).toHaveBeenCalledTimes(5);
+    // Test outputs
```
Comment (jamacku) [inline tests/action.test.ts:223]:
```suggestion

    // Test outputs
```

**Review by jamacku — CHANGES_REQUESTED** [review]

**Review by jamacku — APPROVED** [review]
LGTM, Great work @phracek :tada:

**Pull request merged**